presto-main/src/main/java/io/prestosql/operator/window/WindowPartition.java

```diff
@@ -546,4 +605,201 @@ private long getFrameValue(int channel, String type)
         checkCondition(value >= 0, INVALID_WINDOW_FRAME, "Window frame %s offset must not be negative", value);
         return value;
     }
+
+    private GroupsFrame getFrameRange(FrameInfo frameInfo, GroupsFrame recentFrame)
```

Comment:
There's some duplication and overlapping conditions in this method. How about simplifying to:

```java
private GroupsFrame getFrameRange(FrameInfo frameInfo, GroupsFrame recentFrame)
{
    Type startType = frameInfo.getStartType();
    Type endType = frameInfo.getEndType();

    int start;
    int end;
    int startGroupIndex = GroupsFrame.ignoreIndex();
    int endGroupIndex = GroupsFrame.ignoreIndex();

    switch (startType) {
        case UNBOUNDED_PRECEDING:
            start = 0;
            break;
        case CURRENT_ROW:
            start = peerGroupStart - partitionStart;
            break;
        case PRECEDING: {
            PositionAndGroup frameStart = seek(toIntExact(currentGroupIndex - getStartValue(frameInfo)), recentFrame.getStart(), recentFrame.getStartGroupIndex(), seekGroupStart, () -> new PositionAndGroup(0, 0));
            start = frameStart.getPosition();
            startGroupIndex = frameStart.getGroup();
            break;
        }
        case FOLLOWING: {
            PositionAndGroup frameStart = seek(toIntExact(currentGroupIndex + getStartValue(frameInfo)), recentFrame.getStart(), recentFrame.getStartGroupIndex(), seekGroupStart, () -> new PositionAndGroup(partitionEnd - partitionStart, GroupsFrame.ignoreIndex()));
            start = frameStart.getPosition();
            startGroupIndex = frameStart.getGroup();
            break;
        }
        default:
            throw new UnsupportedOperationException("Unsupported frame start type: " + startType);
    }

    switch (endType) {
        case UNBOUNDED_FOLLOWING:
            end = partitionEnd - partitionStart - 1;
            break;
        case CURRENT_ROW:
            end = peerGroupEnd - partitionStart - 1;
            break;
        case PRECEDING: {
            PositionAndGroup frameEnd = seek(toIntExact(currentGroupIndex - getEndValue(frameInfo)), recentFrame.getEnd(), recentFrame.getEndGroupIndex(), seekGroupEnd, () -> new PositionAndGroup(-1, GroupsFrame.ignoreIndex()));
            end = frameEnd.getPosition();
            endGroupIndex = frameEnd.getGroup();
            break;
        }
        case FOLLOWING: {
            PositionAndGroup frameEnd = seek(toIntExact(currentGroupIndex + getEndValue(frameInfo)), recentFrame.getEnd(), recentFrame.getEndGroupIndex(), seekGroupEnd, () -> new PositionAndGroup(partitionEnd - partitionStart - 1, lastPeerGroup));
            end = frameEnd.getPosition();
            endGroupIndex = frameEnd.getGroup();
            break;
        }
        default:
            throw new UnsupportedOperationException("Unsupported frame end type: " + endType);
    }

    return new GroupsFrame(start, startGroupIndex, end, endGroupIndex);
}
```